sql/inspect: skip hash precheck for non-encodable types

spilchen opened the pull request:
The recent hash precheck optimization for INSPECT (added to speed up index consistency checks) relies on `crdb_internal.datums_to_bytes` to compute row hashes. However, this builtin uses `keyside.Encode` internally, which doesn't support all column types.

When INSPECT encountered tables with TSVECTOR, TSQUERY, or VECTOR (PGVector) columns stored in secondary indexes, the hash precheck would fail with "illegal argument N of type <typename>". This caused INSPECT to fail entirely on these tables.

This commit adds a check to skip the hash precheck if there are any types that cannot be encoded.

Fixes #156175

Epic: CRDB-55075

Release note: none

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/156377)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/inspect/index_consistency_check.go` line 253 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/156377#-Ocl8Job6xD4xAzY75eh:-Ocl8Job6xD4xAzY75ei:bhfixyt) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce5421a5ec38fadd7ca9ee245b785b8e0bb55e80/pkg/sql/inspect/index_consistency_check.go#L253)):*
> ```Go
> 			// all column types (e.g. TSQUERY). So, we look for a specific error
> 			// marker and fallback if found.
> 			if errors.Is(hashErr, builtins.ErrDatumsToBytes_IllegalArg) {
> ```

instead of checking the error after attempting to hash, would it be feasible to check the `types.T` directly?

we use that builtin for hash-sharded indexes, so we already have a check in place for those:
```
root@localhost:26257/defaultdb> create table t (a int primary key, b tsvector, index idx (b) using hash);
ERROR: column b has type tsvector, which is not indexable in a non-inverted index
SQLSTATE: 0A000
```


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/156377)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/inspect/index_consistency_check.go` line 253 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/156377#-Ocl8Job6xD4xAzY75eh:-OclFmBA7ZLcmo73SVsS:buj0vre) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce5421a5ec38fadd7ca9ee245b785b8e0bb55e80/pkg/sql/inspect/index_consistency_check.go#L253)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

instead of checking the error after attempting to hash, would it be feasible to check the `types.T` directly?

we use that builtin for hash-sharded indexes, so we already have a check in place for those:
```
root@localhost:26257/defaultdb> create table t (a int primary key, b tsvector, index idx (b) using hash);
ERROR: column b has type tsvector, which is not indexable in a non-inverted index
SQLSTATE: 0A000
```
</blockquote></details>

This error comes from `ValidateColumnForIndex`, which uses `ColumnTypeIsIndexable`. I could use that helper here, though it means REFCURSOR columns would fall back to the slower join-based inspect instead of using the hash pre-check.

REFCURSOR isn’t allowed in hash indexes, but it can still be serialized with `crdb_internal.datums_to_bytes`:
```
demo@127.0.0.1:26257/demoapp/movr> CREATE TABLE public.refcursor_hash_demo (
                                ->     id  INT8 PRIMARY KEY,
                                ->     cur REFCURSOR
                                -> );
                                ->
                                -> INSERT INTO public.refcursor_hash_demo (id, cur)
                                -> VALUES (1, 'init_cursor');
                                ->

demo@127.0.0.1:26257/demoapp/movr> CREATE INDEX refcursor_hash_demo_cur_idx
                                -> ON public.refcursor_hash_demo (cur)
                                -> USING HASH WITH BUCKET_COUNT = 16;
ERROR: unimplemented: column cur has type refcursor, which is not indexable
SQLSTATE: 0A000
HINT: You have attempted to use a feature that is not yet implemented.
See: https://go.crdb.dev/issue-v/35730/v25.4

demo@127.0.0.1:26257/demoapp/movr> SELECT crdb_internal.datums_to_bytes(cur)
                                -> FROM public.refcursor_hash_demo;
  crdb_internal.datums_to_bytes
----------------------------------
  \x12696e69745f637572736f720001
(1 row)

Time: 1ms total (execution 1ms / network 0ms)

```


If we’re fine with excluding REFCURSOR from the hash pre-check (it’s likely rare), I can use the helper and simplify the code.


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/156377)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/inspect/index_consistency_check.go` line 253 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/156377#-Ocl8Job6xD4xAzY75eh:-OcozML12df_v-ApKfnW:b-bn4x1x) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce5421a5ec38fadd7ca9ee245b785b8e0bb55e80/pkg/sql/inspect/index_consistency_check.go#L253)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

This error comes from `ValidateColumnForIndex`, which uses `ColumnTypeIsIndexable`. I could use that helper here, though it means REFCURSOR columns would fall back to the slower join-based inspect instead of using the hash pre-check.

REFCURSOR isn’t allowed in hash indexes, but it can still be serialized with `crdb_internal.datums_to_bytes`:
```
demo@127.0.0.1:26257/demoapp/movr> CREATE TABLE public.refcursor_hash_demo (
                                ->     id  INT8 PRIMARY KEY,
                                ->     cur REFCURSOR
                                -> );
                                ->
                                -> INSERT INTO public.refcursor_hash_demo (id, cur)
                                -> VALUES (1, 'init_cursor');
                                ->

demo@127.0.0.1:26257/demoapp/movr> CREATE INDEX refcursor_hash_demo_cur_idx
                                -> ON public.refcursor_hash_demo (cur)
                                -> USING HASH WITH BUCKET_COUNT = 16;
ERROR: unimplemented: column cur has type refcursor, which is not indexable
SQLSTATE: 0A000
HINT: You have attempted to use a feature that is not yet implemented.
See: https://go.crdb.dev/issue-v/35730/v25.4

demo@127.0.0.1:26257/demoapp/movr> SELECT crdb_internal.datums_to_bytes(cur)
                                -> FROM public.refcursor_hash_demo;
  crdb_internal.datums_to_bytes
----------------------------------
  \x12696e69745f637572736f720001
(1 row)

Time: 1ms total (execution 1ms / network 0ms)

```


If we’re fine with excluding REFCURSOR from the hash pre-check (it’s likely rare), I can use the helper and simplify the code.
</blockquote></details>

Using #156483 as inspiration, I opted to exclude REFCURSOR. This simplified the code quite a bit. RTAL


<!-- Sent from Reviewable.io -->

rafiss commented on pkg/sql/inspect/index_consistency_check.go:
super nit: is it ever possible for `hashErr` to be nil, but `match` to be false? even if it's not possible now, it might be a good idea to move this log line to be after line 271 (or have a different log message there).

spilchen committed:
sql/inspect: skip hash precheck for non-key-encodable types

The recent hash precheck optimization for INSPECT (added to speed up
index consistency checks) relies on crdb_internal.datums_to_bytes to
compute row hashes. However, this builtin uses keyside.Encode
internally, which doesn't support all column types.

When INSPECT encountered tables with TSVECTOR, TSQUERY, or VECTOR
(PGVector) columns stored in secondary indexes, the hash precheck
would fail with "illegal argument N of type <typename>". This caused
INSPECT to fail entirely on these tables.

This commit adds a check to skip the hash precheck if there are any
types that cannot be encoded.

Fixes #156175

Epic: CRDB-55075

Release note: none

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/156377)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @spilchen)



<!-- Sent from Reviewable.io -->
